File: node/pwire.js
/**
 * Pure-JavaScript PWire binary protocol codec with LZ4 compression
 * and server-side PREPARE/EXECUTE/CLOSE binding.
 *
 * Provides a direct TCP connection to PyroSQL without requiring
 * the FFI shared library (ffi-napi / koffi).
 *
 * @example
 * const { TCPClient } = require('./pwire');
 * const client = new TCPClient('127.0.0.1', 12520);
 * client.connect();
 * const result = client.query('SELECT * FROM users WHERE id = $1', [42]);
 * console.log(result.rows);
 * client.close();
 */

'use strict';

const net = require('net');

// ── LZ4 optional dependency ──────────────────────────────────────────────
let lz4;
try {
    lz4 = require('lz4');
} catch (_) {
    lz4 = null;
}

// ── Protocol constants ───────────────────────────────────────────────────

const MSG_QUERY      = 0x01;
const MSG_PREPARE    = 0x02;
const MSG_EXECUTE    = 0x03;
const MSG_CLOSE      = 0x04;
const MSG_PING       = 0x05;
const MSG_AUTH       = 0x06;
const MSG_COMPRESSED = 0x10;
const MSG_QUIT       = 0xFF;

const RESP_RESULT_SET = 0x01;
const RESP_OK         = 0x02;
const RESP_ERROR      = 0x03;
const RESP_PONG       = 0x04;
const RESP_READY      = 0x05;

const TYPE_NULL  = 0;
const TYPE_I64   = 1;
const TYPE_F64   = 2;
const TYPE_TEXT  = 3;
const TYPE_BOOL  = 4;
const TYPE_BYTES = 5;

const HEADER_SIZE = 5;
const CAP_LZ4 = 0x01;
const COMPRESSION_THRESHOLD = 8 * 1024;

// ── Frame encoding ──────────────────────────────────────────────────────

function frame(type, payload) {
    const buf = Buffer.alloc(HEADER_SIZE + payload.length);
    buf.writeUInt8(type, 0);
    buf.writeUInt32LE(payload.length, 1);
    payload.copy(buf, HEADER_SIZE);
    return buf;
}

function encodeAuth(user, password) {
    const userBuf = Buffer.from(user, 'utf8');
    const passBuf = Buffer.from(password, 'utf8');
    const payload = Buffer.alloc(1 + userBuf.length + 1 + passBuf.length);
    payload.writeUInt8(userBuf.length, 0);
    userBuf.copy(payload, 1);
    payload.writeUInt8(passBuf.length, 1 + userBuf.length);
    passBuf.copy(payload, 2 + userBuf.length);
    return frame(MSG_AUTH, payload);
}

function encodeAuthWithCaps(user, password, caps) {
    const userBuf = Buffer.from(user, 'utf8');
    const passBuf = Buffer.from(password, 'utf8');
    const payload = Buffer.alloc(1 + userBuf.length + 1 + passBuf.length + 1);
    payload.writeUInt8(userBuf.length, 0);
    userBuf.copy(payload, 1);
    payload.writeUInt8(passBuf.length, 1 + userBuf.length);
    passBuf.copy(payload, 2 + userBuf.length);
    payload.writeUInt8(caps, 2 + userBuf.length + passBuf.length);
    return frame(MSG_AUTH, payload);
}

function encodeQuery(sql) {
    return frame(MSG_QUERY, Buffer.from(sql, 'utf8'));
}

function encodePrepare(sql) {
    return frame(MSG_PREPARE, Buffer.from(sql, 'utf8'));
}

function encodeExecute(handle, params) {
    const parts = [];
    const headerBuf = Buffer.alloc(6);
    headerBuf.writeUInt32LE(handle, 0);
    headerBuf.writeUInt16LE(params.length, 4);
    parts.push(headerBuf);
    for (const p of params) {
        const pBuf = Buffer.from(String(p), 'utf8');
        const lenBuf = Buffer.alloc(2);
        lenBuf.writeUInt16LE(pBuf.length, 0);
        parts.push(lenBuf);
        parts.push(pBuf);
    }
    return frame(MSG_EXECUTE, Buffer.concat(parts));
}

function encodeClose(handle) {
    const payload = Buffer.alloc(4);
    payload.writeUInt32LE(handle, 0);
    return frame(MSG_CLOSE, payload);
}

function encodePing() {
    return frame(MSG_PING, Buffer.alloc(0));
}

function encodeQuit() {
    return frame(MSG_QUIT, Buffer.alloc(0));
}

// ── LZ4 compression ─────────────────────────────────────────────────────

/**
 * Compress a frame with LZ4 if beneficial.
 * Uses lz4_flex-compatible format: lz4_data = [u32 LE original_size][raw LZ4 block].
 */
function compressFrame(msgType, payload) {
    if (!lz4 || payload.length <= COMPRESSION_THRESHOLD) {
        return frame(msgType, payload);
    }

    const rawCompressed = Buffer.alloc(lz4.encodeBound(payload.length));
    const rawCompressedSize = lz4.encodeBlock(payload, rawCompressed);
    const rawCompressedBuf = rawCompressed.slice(0, rawCompressedSize);

    // lz4_flex format: prepend u32 LE original size
    const lz4DataLen = 4 + rawCompressedSize;

    const ratio = lz4DataLen / payload.length;
    if (ratio > 0.9) {
        return frame(msgType, payload);
    }

    // Inner: [original_type: u8][uncompressed_length: u32 LE][lz4_data]
    // lz4_data = [u32 LE original_size][raw LZ4 block]
    const inner = Buffer.alloc(1 + 4 + lz4DataLen);
    inner.writeUInt8(msgType, 0);
    inner.writeUInt32LE(payload.length, 1);
    inner.writeUInt32LE(payload.length, 5); // lz4_flex prepended size
    rawCompressedBuf.copy(inner, 9);
    return frame(MSG_COMPRESSED, inner);
}

/**
 * Decompress a MSG_COMPRESSED frame payload.
 * lz4_data is in lz4_flex format: [u32 LE original_size][raw LZ4 block].
 */
function decompressFrame(payload) {
    if (payload.length < 5) {
        throw new Error('Compressed payload too short');
    }
    const originalType = payload.readUInt8(0);
    const uncompressedLen = payload.readUInt32LE(1);
    const lz4Data = payload.slice(5);

    if (lz4Data.length < 4) {
        throw new Error('LZ4 data too short');
    }
    // Skip the 4-byte lz4_flex size prefix
    const rawBlock = lz4Data.slice(4);

    const decompressed = Buffer.alloc(uncompressedLen);
    const decoded = lz4.decodeBlock(rawBlock, decompressed);
    if (decoded < 0) {
        throw new Error('LZ4 decompression failed');
    }
    return { type: originalType, payload: decompressed.slice(0, decoded) };
}

// ── Response decoding ───────────────────────────────────────────────────

function decodeResultSet(payload) {
    let pos = 0;
    const colCount = payload.readUInt16LE(pos); pos += 2;

    const columns = [];
    for (let i = 0; i < colCount; i++) {
        const nameLen = payload.readUInt8(pos); pos++;
        const name = payload.slice(pos, pos + nameLen).toString('utf8'); pos += nameLen;
        const typeTag = payload.readUInt8(pos); pos++;
        columns.push({ name, typeTag });
    }

    const rowCount = payload.readUInt32LE(pos); pos += 4;
    const nullBitmapLen = Math.ceil(colCount / 8);
    const rows = [];

    for (let r = 0; r < rowCount; r++) {
        const bitmap = payload.slice(pos, pos + nullBitmapLen); pos += nullBitmapLen;
        const row = [];
        for (let c = 0; c < colCount; c++) {
            const byteIdx = Math.floor(c / 8);
            const bitIdx = c % 8;
            const isNull = byteIdx < bitmap.length && ((bitmap[byteIdx] >> bitIdx) & 1) === 1;
            if (isNull) {
                row.push(null);
                continue;
            }
            const tt = columns[c].typeTag;
            if (tt === TYPE_I64) {
                // Read as BigInt then convert to Number (may lose precision for very large values)
                const lo = payload.readUInt32LE(pos);
                const hi = payload.readInt32LE(pos + 4);
                row.push(hi * 0x100000000 + lo);
                pos += 8;
            } else if (tt === TYPE_F64) {
                row.push(payload.readDoubleLE(pos));
                pos += 8;
            } else if (tt === TYPE_BOOL) {
                row.push(payload.readUInt8(pos) !== 0);
                pos++;
            } else if (tt === TYPE_BYTES) {
                const len = payload.readUInt16LE(pos); pos += 2;
                row.push(payload.slice(pos, pos + len));
                pos += len;
            } else {
                // TYPE_TEXT and unknown
                const len = payload.readUInt16LE(pos); pos += 2;
                row.push(payload.slice(pos, pos + len).toString('utf8'));
                pos += len;
            }
        }
        rows.push(row);
    }

    return {
        columns: columns.map(c => c.name),
        rows,
        rows_affected: 0
    };
}

function decodeOk(payload) {
    const lo = payload.readUInt32LE(0);
    const hi = payload.readInt32LE(4);
    const rowsAffected = hi * 0x100000000 + lo;
    const tagLen = payload.readUInt8(8);
    const tag = payload.slice(9, 9 + tagLen).toString('utf8');
    return { columns: [], rows: [], rows_affected: rowsAffected, tag };
}

function decodeError(payload) {
    const sqlState = payload.slice(0, 5).toString('ascii');
    const msgLen = payload.readUInt16LE(5);
    const message = payload.slice(7, 7 + msgLen).toString('utf8');
    return { sqlState, message };
}

// ── Synchronous TCP connection ──────────────────────────────────────────

/**
 * Pure-JS synchronous TCP client for PyroSQL.
 * Uses blocking I/O (via synchronous socket reads).
 *
 * Note: Node.js does not natively support synchronous TCP reads.
 * This implementation buffers incoming data and provides a blocking-like API
 * that works with the synchronous FFI test pattern.
 * For production use, prefer the async version or the FFI-based Client.
 */
class TCPClient {
    /**
     * @param {string} host
     * @param {number} port
     * @param {object} [options]
     * @param {string} [options.user='']
     * @param {string} [options.password='']
     * @param {number} [options.timeout=30000]
     */
    constructor(host, port, options = {}) {
        this._host = host;
        this._port = port;
        this._user = options.user || '';
        this._password = options.password || '';
        this._timeout = options.timeout || 30000;
        this._socket = null;
        this._supportsLZ4 = false;
        this._recvBuf = Buffer.alloc(0);
        this._connected = false;
    }

    /**
     * Connect to the server and authenticate.
     */
    connect() {
        return new Promise((resolve, reject) => {
            this._socket = new net.Socket();
            this._socket.setTimeout(this._timeout);

            this._socket.on('data', (data) => {
                this._recvBuf = Buffer.concat([this._recvBuf, data]);
                if (this._pendingResolve && this._tryReadFrame()) {
                    // Frame was read, resolve is handled in _tryReadFrame
                }
            });

            this._socket.on('error', (err) => {
                if (this._pendingReject) {
                    this._pendingReject(err);
                    this._pendingResolve = null;
                    this._pendingReject = null;
                }
            });

            this._socket.connect(this._port, this._host, () => {
                this._connected = true;
                // Authenticate
                const authFrame = lz4
                    ? encodeAuthWithCaps(this._user, this._password, CAP_LZ4)
                    : encodeAuth(this._user, this._password);
                this._socket.write(authFrame);
                this._readFrame().then(({ type, payload }) => {
                    if (type === RESP_ERROR) {
                        const err = decodeError(payload);
                        reject(new Error(`Auth failed: [${err.sqlState}] ${err.message}`));
                        return;
                    }
                    if (type === RESP_READY && payload.length >= 4 && lz4) {
                        const serverCaps = payload.readUInt32LE(0);
                        this._supportsLZ4 = (serverCaps & CAP_LZ4) !== 0;
                    }
                    resolve();
                }).catch(reject);
            });
        });
    }

    /**
     * Execute a query. Uses server-side binding when params are provided.
     * @param {string} sql
     * @param {Array} [params=[]]
     * @returns {Promise<{columns: string[], rows: any[][], rows_affected: number}>}
     */
    async query(sql, params = []) {
        if (params.length > 0) {
            return this._queryPrepared(sql, params);
        }
        this._sendCompressed(MSG_QUERY, Buffer.from(sql, 'utf8'));
        return this._handleResponse();
    }

    /**
     * Execute a DML statement.
     * @param {string} sql
     * @param {Array} [params=[]]
     * @returns {Promise<number>}
     */
    async execute(sql, params = []) {
        if (params.length > 0) {
            const result = await this._queryPrepared(sql, params);
            return result.rows_affected || 0;
        }
        this._sendCompressed(MSG_QUERY, Buffer.from(sql, 'utf8'));
        const result = await this._handleResponse();
        return result.rows_affected || 0;
    }

    /**
     * Server-side prepared statement execution.
     * @private
     */
    async _queryPrepared(sql, params) {
        // PREPARE
        this._sendCompressed(MSG_PREPARE, Buffer.from(sql, 'utf8'));
        const prepFrame = await this._readFrame();
        if (prepFrame.type === RESP_ERROR) {
            const err = decodeError(prepFrame.payload);
            throw new Error(`Prepare error: [${err.sqlState}] ${err.message}`);
        }
        if (prepFrame.type !== RESP_READY || prepFrame.payload.length < 4) {
            throw new Error('Unexpected response for PREPARE');
        }
        const handle = prepFrame.payload.readUInt32LE(0);

        try {
            // EXECUTE
            const wireParams = params.map(v => {
                if (v === null || v === undefined) return 'NULL';
                if (typeof v === 'boolean') return v ? 'true' : 'false';
                return String(v);
            });
            const execFrame = encodeExecute(handle, wireParams);
            const execPayload = execFrame.slice(HEADER_SIZE);
            this._sendCompressed(MSG_EXECUTE, execPayload);
            return await this._handleResponse();
        } finally {
            // CLOSE
            try {
                const closeFrame = encodeClose(handle);
                const closePayload = closeFrame.slice(HEADER_SIZE);
                this._sendCompressed(MSG_CLOSE, closePayload);
                await this._readFrame();
            } catch (_) {
                // best effort
            }
        }
    }

    /**
     * Close the connection.
     */
    close() {
        if (this._socket) {
            try {
                this._socket.write(encodeQuit());
            } catch (_) {}
            this._socket.destroy();
            this._socket = null;
            this._connected = false;
        }
    }

    // ── Internal ─────────────────────────────────────────────────────────

    _sendCompressed(msgType, payload) {
        let data;
        if (this._supportsLZ4) {
            data = compressFrame(msgType, payload);
        } else {
            data = frame(msgType, payload);
        }
        this._socket.write(data);
    }

    _readFrame() {
        return new Promise((resolve, reject) => {
            this._pendingResolve = resolve;
            this._pendingReject = reject;
            // Check if we already have enough data buffered
            this._tryReadFrame();
        });
    }

    _tryReadFrame() {
        if (!this._pendingResolve) return false;
        if (this._recvBuf.length < HEADER_SIZE) return false;

        const msgType = this._recvBuf.readUInt8(0);
        const payloadLen = this._recvBuf.readUInt32LE(1);
        const totalNeeded = HEADER_SIZE + payloadLen;

        if (this._recvBuf.length < totalNeeded) return false;

        const payload = this._recvBuf.slice(HEADER_SIZE, totalNeeded);
        this._recvBuf = this._recvBuf.slice(totalNeeded);

        let result;
        if (msgType === MSG_COMPRESSED && lz4) {
            result = decompressFrame(payload);
        } else {
            result = { type: msgType, payload };
        }

        const resolve = this._pendingResolve;
        this._pendingResolve = null;
        this._pendingReject = null;
        resolve(result);
        return true;
    }

    async _handleResponse() {
        const { type, payload } = await this._readFrame();
        if (type === RESP_RESULT_SET) {
            return decodeResultSet(payload);
        } else if (type === RESP_OK) {
            return decodeOk(payload);
        } else if (type === RESP_ERROR) {
            const err = decodeError(payload);
            throw new Error(`[${err.sqlState}] ${err.message}`);
        } else {
            throw new Error(`Unexpected response type 0x${type.toString(16)}`);
        }
    }
}

// ── Exports ─────────────────────────────────────────────────────────────

module.exports = {
    // Protocol constants
    MSG_QUERY, MSG_PREPARE, MSG_EXECUTE, MSG_CLOSE, MSG_PING, MSG_AUTH,
    MSG_COMPRESSED, MSG_QUIT,
    RESP_RESULT_SET, RESP_OK, RESP_ERROR, RESP_PONG, RESP_READY,
    TYPE_NULL, TYPE_I64, TYPE_F64, TYPE_TEXT, TYPE_BOOL, TYPE_BYTES,
    HEADER_SIZE, CAP_LZ4, COMPRESSION_THRESHOLD,

    // Encoding
    frame, encodeAuth, encodeAuthWithCaps, encodeQuery, encodePrepare,
    encodeExecute, encodeClose, encodePing, encodeQuit,

    // Compression
    compressFrame, decompressFrame,

    // Decoding
    decodeResultSet, decodeOk, decodeError,

    // Client
    TCPClient,
};

File: python/pyrosql/pwire.py
"""Pure-Python PWire binary protocol codec with LZ4 compression support.

This module provides a direct TCP connection to PyroSQL without requiring
the FFI shared library, including LZ4 compression and server-side binding
via PREPARE/EXECUTE/CLOSE messages.
"""

from __future__ import annotations

import socket
import struct
from typing import Any, Dict, List, Optional, Sequence, Tuple

try:
    import lz4.block as lz4_block
    _HAS_LZ4 = True
except ImportError:
    _HAS_LZ4 = False


# Message types (client -> server)
MSG_QUERY = 0x01
MSG_PREPARE = 0x02
MSG_EXECUTE = 0x03
MSG_CLOSE = 0x04
MSG_PING = 0x05
MSG_AUTH = 0x06
MSG_COMPRESSED = 0x10
MSG_QUIT = 0xFF

# Response types (server -> client)
RESP_RESULT_SET = 0x01
RESP_OK = 0x02
RESP_ERROR = 0x03
RESP_PONG = 0x04
RESP_READY = 0x05

# Value type tags
TYPE_NULL = 0
TYPE_I64 = 1
TYPE_F64 = 2
TYPE_TEXT = 3
TYPE_BOOL = 4
TYPE_BYTES = 5

HEADER_SIZE = 5
CAP_LZ4 = 0x01
COMPRESSION_THRESHOLD = 8 * 1024


def _frame(msg_type: int, payload: bytes) -> bytes:
    """Build a PWire frame: 1 byte type + 4 byte LE length + payload."""
    return struct.pack("<BI", msg_type, len(payload)) + payload


def encode_auth(user: str, password: str) -> bytes:
    user_b = user.encode("utf-8")
    pass_b = password.encode("utf-8")
    payload = struct.pack("B", len(user_b)) + user_b + struct.pack("B", len(pass_b)) + pass_b
    return _frame(MSG_AUTH, payload)


def encode_auth_with_caps(user: str, password: str, caps: int) -> bytes:
    user_b = user.encode("utf-8")
    pass_b = password.encode("utf-8")
    payload = (
        struct.pack("B", len(user_b)) + user_b
        + struct.pack("B", len(pass_b)) + pass_b
        + struct.pack("B", caps)
    )
    return _frame(MSG_AUTH, payload)


def encode_query(sql: str) -> bytes:
    return _frame(MSG_QUERY, sql.encode("utf-8"))


def encode_prepare(sql: str) -> bytes:
    return _frame(MSG_PREPARE, sql.encode("utf-8"))


def encode_execute(handle: int, params: List[str]) -> bytes:
    payload = struct.pack("<IH", handle, len(params))
    for p in params:
        p_bytes = p.encode("utf-8")
        payload += struct.pack("<H", len(p_bytes)) + p_bytes
    return _frame(MSG_EXECUTE, payload)


def encode_close(handle: int) -> bytes:
    return _frame(MSG_CLOSE, struct.pack("<I", handle))


def encode_ping() -> bytes:
    return _frame(MSG_PING, b"")


def encode_quit() -> bytes:
    return _frame(MSG_QUIT, b"")


def compress_frame(msg_type: int, payload: bytes) -> bytes:
    """Compress a frame with LZ4 if beneficial. Returns framed bytes.

    Uses lz4_flex-compatible format: the lz4_data within the inner frame is
    [u32 LE original_size][raw LZ4 block].
    """
    if not _HAS_LZ4 or len(payload) <= COMPRESSION_THRESHOLD:
        return _frame(msg_type, payload)

    raw_compressed = lz4_block.compress(payload, store_size=False)
    # lz4_flex format: prepend u32 LE original size
    lz4_data = struct.pack("<I", len(payload)) + raw_compressed

    ratio = len(lz4_data) / len(payload)
    if ratio > 0.9:
        return _frame(msg_type, payload)

    # Inner: [original_type: u8][uncompressed_length: u32 LE][lz4_data]
    inner = struct.pack("<BI", msg_type, len(payload)) + lz4_data
    return _frame(MSG_COMPRESSED, inner)


def decompress_frame(payload: bytes) -> Tuple[int, bytes]:
    """Decompress a MSG_COMPRESSED payload. Returns (original_type, data).

    lz4_data is in lz4_flex format: [u32 LE original_size][raw LZ4 block].
    """
    if len(payload) < 5:
        raise ValueError("Compressed payload too short")

    original_type = payload[0]
    uncompressed_len = struct.unpack_from("<I", payload, 1)[0]
    lz4_data = payload[5:]

    # lz4_data starts with [u32 LE original size] from lz4_flex, skip it
    if len(lz4_data) < 4:
        raise ValueError("LZ4 data too short")
    raw_block = lz4_data[4:]
    decompressed = lz4_block.decompress(raw_block, uncompressed_size=uncompressed_len)
    return original_type, decompressed


def _recv_exact(sock: socket.socket, count: int) -> bytes:
    """Read exactly count bytes from a socket."""
    buf = bytearray()
    while len(buf) < count:
        chunk = sock.recv(count - len(buf))
        if not chunk:
            raise ConnectionError("Connection closed by server")
        buf.extend(chunk)
    return bytes(buf)


def read_frame(sock: socket.socket) -> Tuple[int, bytes]:
    """Read a complete frame from a socket. Transparently decompresses MSG_COMPRESSED."""
    header = _recv_exact(sock, HEADER_SIZE)
    msg_type, length = struct.unpack("<BI", header)

    payload = _recv_exact(sock, length) if length > 0 else b""

    if msg_type == MSG_COMPRESSED:
        orig_type, decompressed = decompress_frame(payload)
        return orig_type, decompressed

    return msg_type, payload


def decode_result_set(payload: bytes) -> Dict[str, Any]:
    """Decode a RESP_RESULT_SET payload into columns and rows."""
    pos = 0
    col_count = struct.unpack_from("<H", payload, pos)[0]
    pos += 2

    columns = []
    for _ in range(col_count):
        name_len = payload[pos]
        pos += 1
        name = payload[pos:pos + name_len].decode("utf-8")
        pos += name_len
        type_tag = payload[pos]
        pos += 1
        columns.append({"name": name, "type_tag": type_tag})

    row_count = struct.unpack_from("<I", payload, pos)[0]
    pos += 4
    null_bitmap_len = (col_count + 7) // 8

    rows: List[List[Any]] = []
    for _ in range(row_count):
        bitmap = payload[pos:pos + null_bitmap_len]
        pos += null_bitmap_len
        row: List[Any] = []
        for c in range(col_count):
            byte_idx = c // 8
            bit_idx = c % 8
            is_null = byte_idx < len(bitmap) and ((bitmap[byte_idx] >> bit_idx) & 1) == 1
            if is_null:
                row.append(None)
                continue
            tt = columns[c]["type_tag"]
            if tt == TYPE_I64:
                val = struct.unpack_from("<q", payload, pos)[0]
                pos += 8
                row.append(val)
            elif tt == TYPE_F64:
                val = struct.unpack_from("<d", payload, pos)[0]
                pos += 8
                row.append(val)
            elif tt == TYPE_BOOL:
                val = payload[pos] != 0
                pos += 1
                row.append(val)
            elif tt == TYPE_BYTES:
                l = struct.unpack_from("<H", payload, pos)[0]
                pos += 2
                row.append(payload[pos:pos + l])
                pos += l
            else:
                # TYPE_TEXT and unknown
                l = struct.unpack_from("<H", payload, pos)[0]
                pos += 2
                row.append(payload[pos:pos + l].decode("utf-8"))
                pos += l
        rows.append(row)

    return {
        "columns": [c["name"] for c in columns],
        "rows": rows,
        "rows_affected": 0,
    }


def decode_ok(payload: bytes) -> Dict[str, Any]:
    """Decode a RESP_OK payload."""
    rows_affected = struct.unpack_from("<q", payload, 0)[0]
    tag_len = payload[8]
    tag = payload[9:9 + tag_len].decode("utf-8")
    return {"columns": [], "rows": [], "rows_affected": rows_affected, "tag": tag}


def decode_error(payload: bytes) -> Tuple[str, str]:
    """Decode a RESP_ERROR payload. Returns (sql_state, message)."""
    sql_state = payload[0:5].decode("ascii")
    msg_len = struct.unpack_from("<H", payload, 5)[0]
    message = payload[7:7 + msg_len].decode("utf-8")
    return sql_state, message


class PWireConnection:
    """Pure-Python TCP connection to PyroSQL with LZ4 and server-side binding."""

    def __init__(self, host: str, port: int, timeout: float = 30.0):
        self._sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        self._sock.settimeout(timeout)
        self._sock.setsockopt(socket.IPPROTO_TCP, socket.TCP_NODELAY, 1)
        self._sock.connect((host, port))
        self._supports_lz4 = False

    def authenticate(self, user: str = "", password: str = "") -> None:
        if _HAS_LZ4:
            self._send_raw(encode_auth_with_caps(user, password, CAP_LZ4))
        else:
            self._send_raw(encode_auth(user, password))

        msg_type, payload = read_frame(self._sock)

        if msg_type == RESP_ERROR:
            sql_state, message = decode_error(payload)
            raise ConnectionError(f"[{sql_state}] {message}")

        if msg_type not in (RESP_READY, RESP_OK):
            raise ConnectionError(f"Unexpected auth response type 0x{msg_type:02X}")

        if msg_type == RESP_READY and len(payload) >= 4 and _HAS_LZ4:
            server_caps = struct.unpack_from("<I", payload, 0)[0]
            self._supports_lz4 = (server_caps & CAP_LZ4) != 0

    def query(self, sql: str) -> Dict[str, Any]:
        self._send_compressed(MSG_QUERY, sql.encode("utf-8"))
        return self._handle_response()

    def prepare(self, sql: str) -> int:
        """Send MSG_PREPARE. Returns the u32 handle from RESP_READY."""
        self._send_compressed(MSG_PREPARE, sql.encode("utf-8"))
        msg_type, payload = read_frame(self._sock)
        if msg_type == RESP_ERROR:
            sql_state, message = decode_error(payload)
            raise QueryError(f"[{sql_state}] {message}")
        if msg_type != RESP_READY:
            raise QueryError(f"Expected READY from PREPARE, got 0x{msg_type:02X}")
        if len(payload) < 4:
            raise QueryError("PREPARE response too short")
        return struct.unpack_from("<I", payload, 0)[0]

    def execute_prepared(self, handle: int, params: List[str]) -> Dict[str, Any]:
        """Send MSG_EXECUTE with handle + params and return the response."""
        full_frame = encode_execute(handle, params)
        exec_payload = full_frame[HEADER_SIZE:]
        self._send_compressed(MSG_EXECUTE, exec_payload)
        return self._handle_response()

    def close_prepared(self, handle: int) -> None:
        """Send MSG_CLOSE for the given handle."""
        full_frame = encode_close(handle)
        close_payload = full_frame[HEADER_SIZE:]
        self._send_compressed(MSG_CLOSE, close_payload)
        msg_type, payload = read_frame(self._sock)
        if msg_type == RESP_ERROR:
            sql_state, message = decode_error(payload)
            raise QueryError(f"[{sql_state}] {message}")

    def execute(self, sql: str) -> Dict[str, Any]:
        self._send_compressed(MSG_QUERY, sql.encode("utf-8"))
        return self._handle_response()

    def ping(self) -> bool:
        self._send_compressed(MSG_PING, b"")
        msg_type, _ = read_frame(self._sock)
        return msg_type == RESP_PONG

    def close(self) -> None:
        try:
            self._send_raw(encode_quit())
        except Exception:
            pass
        try:
            self._sock.close()
        except Exception:
            pass

    def _send_raw(self, data: bytes) -> None:
        self._sock.sendall(data)

    def _send_compressed(self, msg_type: int, payload: bytes) -> None:
        if self._supports_lz4:
            data = compress_frame(msg_type, payload)
        else:
            data = _frame(msg_type, payload)
        self._send_raw(data)

    def _handle_response(self) -> Dict[str, Any]:
        msg_type, payload = read_frame(self._sock)
        if msg_type == RESP_RESULT_SET:
            return decode_result_set(payload)
        elif msg_type == RESP_OK:
            return decode_ok(payload)
        elif msg_type == RESP_ERROR:
            sql_state, message = decode_error(payload)
            raise QueryError(f"[{sql_state}] {message}")
        else:
            raise QueryError(f"Unexpected response type 0x{msg_type:02X}")


class QueryError(Exception):
    """Raised when a query fails."""
